File: 21. Merge Two Sorted Lists/CodeVersion2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        newList = ListNode()
        p = newList

        while list1 or list2:
            if not list1:
                p.next = ListNode(list2.val)
                p = p.next
                list2 = list2.next
            elif not list2:
                p.next = ListNode(list1.val)
                p = p.next
                list1 = list1.next
            elif list1.val < list2.val:
                p.next = ListNode(list1.val)
                p = p.next
                list1 = list1.next
            else:
                p.next = ListNode(list2.val)
                p = p.next
                list2 = list2.next
                
        return newList.next

File: 21. Merge Two Sorted Lists/CodeVersion2.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} list1
 * @param {ListNode} list2
 * @return {ListNode}
 */
var mergeTwoLists = function(list1, list2) {
    let newList = new ListNode();
    let p = newList;

    while (list1 !== null || list2 !== null) {
        if (list1 === null) {
            p.next = new ListNode(list2.val);
            p = p.next;
            list2 = list2.next;
        } else if (list2 === null) {
            p.next = new ListNode(list1.val);
            p = p.next;
            list1 = list1.next;
        } else if (list1.val < list2.val) {
            p.next = new ListNode(list1.val);
            p = p.next;
            list1 = list1.next;
        } else {
            p.next = new ListNode(list2.val);
            p = p.next;
            list2 = list2.next;
        }
    }

    return newList.next;
    
};
